crates/cext/src/sparse_observable.rs

```diff
@@ -0,0 +1,749 @@
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_obs_term(
+    obs: *mut SparseObservable,
+    index: u64,
+    out: *mut CSparseTerm,
+) -> ExitCode {
+    let out = unsafe { mut_ptr_as_ref(out) };
+    let obs = unsafe { mut_ptr_as_ref(obs) };
+
+    let index = index as usize;
+    if index > obs.num_terms() {
+        return ExitCode::IndexError;
+    }
+
+    out.len = obs.boundaries()[index + 1] - obs.boundaries()[index];
+    out.coeff = &mut obs.coeffs_mut()[index];
+    out.num_qubits = obs.num_qubits();
+
+    let start = obs.boundaries()[index];
+    out.bit_terms = &mut obs.bit_terms_mut()[start];
+    out.indices = unsafe { &mut obs.indices_mut()[start] };
+
+    ExitCode::Success
+}
+
+/// @ingroup QkSparseObservable
+/// Get the number of terms in the observable.
+///
+/// @param obs A pointer to the observable.
+///
+/// @return The number of terms in the observable.
+///
+/// # Example
+///
+///     QkSparseObservable *obs = qk_obs_identity(100);
+///     uint64_t num_terms = qk_obs_num_terms(obs);  // num_terms==1
+///
+/// # Safety
+///
+/// Behavior is undefined ``obs`` is not a valid, non-null pointer to a ``QkSparseObservable``.
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_obs_num_terms(obs: *const SparseObservable) -> u64 {
+    let obs = unsafe { const_ptr_as_ref(obs) };
+    obs.num_terms() as u64
+}
+
+/// @ingroup QkSparseObservable
+/// Get the number of qubits the observable is defined on.
+///
+/// @param obs A pointer to the observable.
+///
+/// @return The number of qubits the observable is defined on.
+///
+/// # Example
+///
+///     QkSparseObservable *obs = qk_obs_identity(100);
+///     uint32_t num_qubits = qk_obs_num_qubits(obs);  // num_qubits==100
+///
+/// # Safety
+///
+/// Behavior is undefined ``obs`` is not a valid, non-null pointer to a ``QkSparseObservable``.
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_obs_num_qubits(obs: *const SparseObservable) -> u32 {
+    let obs = unsafe { const_ptr_as_ref(obs) };
+    obs.num_qubits()
+}
+
+/// @ingroup QkSparseObservable
+/// Get the number of bit terms/indices in the observable.
+///
+/// @param obs A pointer to the observable.
+///
+/// @return The number of terms in the observable.
+///
+/// # Example
+///
+///     QkSparseObservable *obs = qk_obs_identity(100);
+///     uint64_t len = qk_obs_len(obs);  // len==0, as there are no non-trivial bit terms
+///
+/// # Safety
+///
+/// Behavior is undefined ``obs`` is not a valid, non-null pointer to a ``QkSparseObservable``.
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_obs_len(obs: *const SparseObservable) -> u64 {
+    let obs = unsafe { const_ptr_as_ref(obs) };
+    obs.bit_terms().len() as u64
+}
+
+/// @ingroup QkSparseObservable
+/// Get a pointer to the coefficients.
+///
+/// This can be used to read and modify the observable's coefficients.
```

Comment:
Would be good to comment about the data-length validity of the pointer (here and for other similar methods).

Reply:
Added in https://github.com/Qiskit/qiskit/commit/bc376c6f27b20c932c7eb4e3da7e63ffb9cbf2ea, let me know if this is what you had in mind 🙂